partiql-logical-planner/src/lower.rs

```diff
@@ -543,118 +613,192 @@ impl<'ast> Visitor<'ast> for AstToLogical {
+                return Recurse::Stop;
+            }
         }
+        Recurse::Continue
     }
 
-    fn enter_set_expr(&mut self, _set_expr: &'ast SetExpr) {}
+    fn enter_set_expr(&mut self, _set_expr: &'ast SetExpr) -> Recurse {
+        Recurse::Continue
+    }
 
-    fn exit_set_expr(&mut self, _set_expr: &'ast SetExpr) {}
+    fn exit_set_expr(&mut self, _set_expr: &'ast SetExpr) -> Recurse {
+        Recurse::Continue
+    }
 
-    fn enter_select(&mut self, _select: &'ast Select) {}
+    fn enter_select(&mut self, _select: &'ast Select) -> Recurse {
+        Recurse::Continue
+    }
 
-    fn exit_select(&mut self, _select: &'ast Select) {}
+    fn exit_select(&mut self, _select: &'ast Select) -> Recurse {
+        Recurse::Continue
+    }
 
-    fn enter_projection(&mut self, _projection: &'ast Projection) {
+    fn enter_projection(&mut self, _projection: &'ast Projection) -> Recurse {
         self.enter_benv();
         self.enter_env();
+        Recurse::Continue
     }
 
-    fn exit_projection(&mut self, _projection: &'ast Projection) {
+    fn exit_projection(&mut self, _projection: &'ast Projection) -> Recurse {
         let benv = self.exit_benv();
-        assert_eq!(benv.len(), 0);
+        if self.eq_or_error(benv.len(), 0, "benv.len() != 0") == Recurse::Stop {
+            return Recurse::Stop;
+        }
+
         let env = self.exit_env();
-        assert_eq!(env.len(), 0);
+        if self.eq_or_error(env.len(), 0, "env.len() != 0") == Recurse::Stop {
+            return Recurse::Stop;
+        }
 
         if let Some(SetQuantifier::Distinct) = _projection.setq {
             let id = self.plan.add_operator(BindingsOp::Distinct);
             self.current_clauses_mut().distinct.replace(id);
         }
+        Recurse::Continue
     }
 
-    fn enter_projection_kind(&mut self, _projection_kind: &'ast ProjectionKind) {
+    fn enter_projection_kind(&mut self, _projection_kind: &'ast ProjectionKind) -> Recurse {
         self.enter_benv();
         self.enter_env();
+        Recurse::Continue
     }
 
-    fn exit_projection_kind(&mut self, _projection_kind: &'ast ProjectionKind) {
+    fn exit_projection_kind(&mut self, _projection_kind: &'ast ProjectionKind) -> Recurse {
         let benv = self.exit_benv();
-        assert_eq!(benv.len(), 0); // TODO sub-query
+        if !benv.is_empty() {
+            self.errors.push(LowerError::NotYetImplemented(
+                "Subquery within project".to_string(),
+            ));
+            return Recurse::Stop;
+        }
         let env = self.exit_env();
 
         let select: BindingsOp = match _projection_kind {
             ProjectionKind::ProjectStar => logical::BindingsOp::ProjectAll,
             ProjectionKind::ProjectList(_) => {
+                if self.true_or_error(env.len().is_even(), "env.len() is not even") == Recurse::Stop
+                {
+                    return Recurse::Stop;
+                }
```

Comment:
It may be worth considering turning `true_or_error` and `eq_or_error` into macros, as a macro could return from the current `fn` without needing the `if ... = Stop { return Stop; }` pattern.  

You may need both a 'critical' and a 'recoverable' version of each; with the 'critical' one immediately returning `Stop`, and the 'recoverable' just pushing to the `errors` state. Having 'critical_err' vs 'recoverable_err' (or something like that) in the name would also go a long way towards making it clear what errors are 'stop the world' and which we allow lowering to continue, if only to build up more error state; This is accomplished now with comments like `// Report error but allow visitor to continue` or `// Include as an error but allow lowering to proceed for multiple error reporting`

Reply:
Agree on the clarity that the macro would provide. Almost ended up creating them but wasn't sure if it was overkill. I'll change these call sites to use a macro in an upcoming commit.